midp/gfx.js

```diff
@@ -357,20 +479,19 @@ var currentlyFocusedTextEditor;
         this.baseline = size | 0;
         this.height = (size * 1.3) | 0;
 
-        this.context = document.createElement("canvas").getContext("2d");
-        this.context.canvas.width = 0;
-        this.context.canvas.height = 0;
-        this.context.font = style + size + "px " + face;
         this.size = size;
+        // XXX Figure out why this works when these two fields are both int.
         this.style = style;
         this.face = face;
     };
 
-    function calcStringWidth(font, str) {
+    function calcStringWidth(font, str, scaleByDPR) {
         var emojiLen = 0;
+        var fontSize = font.size * (scaleByDPR ? DPR : 1);
 
-        var len = font.context.measureText(str.replace(emoji.regEx, function() {
-            emojiLen += font.size;
+        tempContext.font = font.style + " " + fontSize + "px " + font.face;
+        var len = tempContext.measureText(str.replace(emoji.regEx, function() {
+            emojiLen += fontSize;
```

Comment:
> I didn't notice one, but I just ran GraphicsBench, and it reports that "Graphics.drawString baseline centered" and "Offscreen Graphics.drawString baseline centered" are roughly half as fast on this branch, regardless of DPR. So this branch does indeed regress string drawing perf. I'll look into it.

291516a claws back most of the perf hit. It creates two contexts for measuring string widths, one for measuring unscaled widths and the other for measuring scaled widths (with an optimization to create only a single context if DPR == 1).

On my machine, the Graphics.drawString benchmarks perform 42-43K ops per second on master. On this branch, they now hit 34-35K. I think the rest is the necessary cost of drawing larger strings to a larger canvas when DPR == 2 (as it is on my machine).

Note that the other graphics benchmark, Graphics.drawRegion, also performs worse on this branch. That's expected since scaling the canvas means drawing more pixels, up to four times more when DPR == 2.

Regarding the use of two contexts to measure the scaled vs. unscaled widths of strings, I did some experimentation to confirm my earlier findings that merely multiplying an unscaled width by the DPR didn't give me an accurate scaled width.

Here are some measurements for "The quick brown fox jumps over the lazy dog." in two font styles (regular and "bold") at both 9px and 18px (per a theoretical DPR of 2). I calculate the result of simply multiplying the width at 9px by the theoretical DPR, and I also calculate what the return value of calcStringWidth would be using the current integer conversion mechanism (bitwise or) and a prospective alternative (Math.round).

9px sans-serif:
* 180.23333740234375
* 180.23333740234375 * 2 = 360.4666748046875
* 180.23333740234375 * 2 | 0 = 360
* Math.round(180.23333740234375) = 360

18px sans-serif:
* 360.73333740234375
* 360.73333740234375 | 0 = 360
* Math.round(360.73333740234375) = 361

bold 9px sans-serif:
* 194.43333435058594
* 194.43333435058594 * 2 = 388.8666687011719
* 194.43333435058594 * 2 | 0 = 388
* Math.round(194.43333435058594 * 2) = 389

bold 18px sans-serif:
* 389.4333190917969
* 389.4333190917969 | 0 = 389
* Math.round(389.4333190917969 | 0) = 389

Multiplying the unscaled widths by 2 gets close, but it can be a pixel off, regardless of the integer conversion. Perhaps that's ok, and we should simplify the scaled measurement. Unsure.